kernel/src/process_loading.rs

```diff
@@ -0,0 +1,769 @@
+    fault_policy: &'static dyn ProcessFaultPolicy,
+    capability: &dyn ProcessManagementCapability,
+) -> Result<(), ProcessLoadError> {
+    if config::CONFIG.debug_load_processes {
+        debug!(
+            "Loading processes from flash={:#010X}-{:#010X} into sram={:#010X}-{:#010X}",
+            app_flash.as_ptr() as usize,
+            app_flash.as_ptr() as usize + app_flash.len() - 1,
+            app_memory.as_ptr() as usize,
+            app_memory.as_ptr() as usize + app_memory.len() - 1
+        );
+    }
+
+    let mut remaining_flash = app_flash;
+    let mut remaining_memory = app_memory;
+    // Try to discover up to `procs.len()` processes in flash.
+    let mut index = 0;
+    let num_procs = procs.len();
+    while index < num_procs {
+        let load_result = load_process(
+            kernel,
+            chip,
+            remaining_flash,
+            remaining_memory,
+            index,
+            fault_policy,
+            capability,
+        );
+        match load_result {
+            Ok((new_flash, new_mem, proc)) => {
+                remaining_flash = new_flash;
+                remaining_memory = new_mem;
+                if proc.is_some() {
+                    if config::CONFIG.debug_load_processes {
+                        proc.map(|p| debug!("Loaded process {}", p.get_process_name()));
+                    }
+                    procs[index] = proc;
+                    index = index + 1;
+                }
+            }
+            Err((_new_flash, _new_mem, err)) => {
+                if config::CONFIG.debug_load_processes {
+                    debug!("No more processes to load: {:?}.", err);
+                }
+                // No more processes to load.
+                break;
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Use `checker` to transition `procs` from the `Unverified` into the
+/// `Unstarted` state (if they pass the checker policy) or
+/// `CredentialsFailed` state (if they do not pass the checker
+/// policy). If `verifier` is `None` then all processes are
+/// automatically verified. Processes that transition into the
+/// `Unstarted` state are started by enqueueing a stack frame to run
+/// the initialization function as indicated in the TBF header.
+#[inline(always)]
+fn check_processes<'a, KR: KernelResources<C>, C: Chip>(
+    procs: &'static [Option<&'static dyn Process>],
+    kernel_resources: &KR,
+) -> Result<(), ProcessLoadError>
+//    where <KR as KernelResources<C>>::CredentialsCheckingPolicy: 'static,
+{
+    let policy = kernel_resources.credentials_checking_policy();
+    #[allow(unused_mut)]
+    let machine = unsafe {
+        static_init!(
+            ProcessCheckerMachine,
+            ProcessCheckerMachine {
+                process: Cell::new(0),
+                footer: Cell::new(0),
+                checker: OptionalCell::empty(),
+                processes: procs,
+            }
+        )
+    };
```

Comment:
I'm still uncomfortable with this here. We currently only use static_init from the board main.rs and not in the kernel. Including this here, as unsafe as well, sets a precedent I don't think we want to.

Reply:
I agree, it's not great. My reasoning is that the fact that `load_and_check_processes` creates a checking state machine is an internal implementation detail. So it's something that should not be leaked to `main.rs` and a board file. The kernel should be able to change this without having to change every board.

Generally speaking, the state machine will typically need a static lifetime, as it depends checkers with static lifetimes, as they depend on cryptographic HILs which use static lifetimes. 

One approach might be to instantiate it in the `Kernel` struct? We will need to be careful that platforms which don't engage the checking state machine don't include its code.

Reply:
note: the `unsafe` will go away with #3239.

Reply:
Following the discussion we had the recent OT call, I moved allocation of the ProcessCheckerMachine into the `Kernel` struct, and added a method to access it.